pkg/controller/certificatesigningrequests/venafi/venafi.go

```diff
@@ -0,0 +1,206 @@
+	log := logf.FromContext(ctx, "sign")
+	log = logf.WithRelatedResource(log, issuerObj)
+
+	resourceNamespace := v.issuerOptions.ResourceNamespace(issuerObj)
+
+	client, err := v.clientBuilder(resourceNamespace, v.secretsLister, issuerObj)
+	if apierrors.IsNotFound(err) {
+		message := "Required secret resource not found"
+		v.recorder.Event(csr, corev1.EventTypeWarning, "SecretNotFound", message)
+		log.Error(err, message)
+		return nil
+	}
+
+	if err != nil {
+		message := fmt.Sprintf("Failed to initialise venafi client for signing: %s", err)
+		v.recorder.Event(csr, corev1.EventTypeWarning, "ErrorVenafiInit", message)
+		log.Error(err, message)
+		return err
+	}
+
+	var customFields []venafiapi.CustomField
+	if annotation, exists := csr.GetAnnotations()[experimentalapi.CertificateSigningRequestVenafiCustomFieldsAnnotationKey]; exists && annotation != "" {
+		err := json.Unmarshal([]byte(annotation), &customFields)
+		if err != nil {
+			message := fmt.Sprintf("Failed to parse %q annotation: %s", experimentalapi.CertificateSigningRequestVenafiCustomFieldsAnnotationKey, err)
+			v.recorder.Event(csr, corev1.EventTypeWarning, "ErrorCustomFields", message)
+			util.CertificateSigningRequestSetFailed(csr, "ErrorCustomFields", message)
+			_, err = v.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+			return err
+		}
+	}
+
+	duration, err := pki.DurationFromCertificateSigningRequest(csr)
+	if err != nil {
+		message := fmt.Sprintf("Failed to parse requested duration: %s", err)
+		log.Error(err, message)
+		v.recorder.Event(csr, corev1.EventTypeWarning, "ErrorParseDuration", message)
+		util.CertificateSigningRequestSetFailed(csr, "ErrorParseDuration", message)
+		_, err := v.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+		return err
+	}
+
+	// The signing process with Venafi is slow. The "pickupID" allows us to track
+	// the progress of the certificate signing. It is set as an annotation the
+	// first time the Certificate is reconciled.
+	pickupID := csr.GetAnnotations()[experimentalapi.CertificateSigningRequestVenafiPickupIDAnnotationKey]
+
+	// check if the pickup ID annotation is there, if not set it up.
+	if len(pickupID) == 0 {
+		pickupID, err := client.RequestCertificate(csr.Spec.Request, duration, customFields)
+		// Check some known error types
+		if err != nil {
+			switch err.(type) {
+
+			case venaficlient.ErrCustomFieldsType:
+				log.Error(err, "")
+				v.recorder.Event(csr, corev1.EventTypeWarning, "ErrorCustomFields", err.Error())
+				util.CertificateSigningRequestSetFailed(csr, "ErrorCustomFields", err.Error())
+				_, err := v.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+				return err
+
+			default:
+				message := fmt.Sprintf("Failed to request venafi certificate: %s", err)
+				log.Error(err, message)
+				v.recorder.Event(csr, corev1.EventTypeWarning, "ErrorRequest", message)
+				util.CertificateSigningRequestSetFailed(csr, "ErrorRequest", message)
```

Comment:
question: is this function mutating the `csr`? If it does, shouldn't we do a deep copy? Or is it already a deep copy?

It would be nice to have this information somewhere (e.g., at the top of this func), since I never know:

1. Is it a reference to an object in cache like what we do in the rest of cert-manager?
2. Or does it come from controller-runtime which means it is a deep copy?

Reply:
I have added a `DeepCopy` to the stop `Sync` func of the shared controller so there is no doubt.

Reply:
Thanks!

Just to add to the conversation, I looked at the rest of cert-manager and are not consistent with "when" we do the deep copy, and we never detail anywhere "why" we do one or the other:

|                     File                      | Upfront | Last moment |
|-----------------------------------------------|---------|-------------|
| [`acmechallenges/sync.go`][]                  | x       |             |
| [`acmeorders/sync.go`][]                      | x       |             |
| [`clusterissuers/sync.go`][]                  | x       |             |
| [`issuers/sync.go`][]                         | x       |             |
| [`acmechallenges/sync.go`][]                  | x       |             |
| [`certificaterequests/sync.go`][]             | x       |             |
| `issuing_controller.go` ([1][], [2][], [3][]) |         | x           |
| [`keymanager_controller.go`][]                |         | x           |
| [`readiness_controller.go`][]                 |         | x           |
| [`trigger_controller.go`][]                   |         | x           |

[`acmechallenges/sync.go`]: https://github.com/jetstack/cert-manager/blob/88e85d07253f8d8a99628a2220c7f751a23ef4f1/pkg/controller/acmechallenges/sync.go#L68
[`acmeorders/sync.go`]: https://github.com/jetstack/cert-manager/blob/88e85d07253f8d8a99628a2220c7f751a23ef4f1/pkg/controller/acmeorders/sync.go#L60
[1]: https://github.com/jetstack/cert-manager/blob/88e85d07253f8d8a99628a2220c7f751a23ef4f1/pkg/controller/certificates/issuing/issuing_controller.go#L311
[2]: https://github.com/jetstack/cert-manager/blob/88e85d07253f8d8a99628a2220c7f751a23ef4f1/pkg/controller/certificates/issuing/issuing_controller.go#L328
[3]: https://github.com/jetstack/cert-manager/blob/88e85d07253f8d8a99628a2220c7f751a23ef4f1/pkg/controller/certificates/issuing/temporary.go#L45
[`keymanager_controller.go`]: https://github.com/jetstack/cert-manager/blob/88e85d07253f8d8a99628a2220c7f751a23ef4f1/pkg/controller/certificates/keymanager/keymanager_controller.go#L304
[`readiness_controller.go`]: https://github.com/jetstack/cert-manager/blob/88e85d07253f8d8a99628a2220c7f751a23ef4f1/pkg/controller/certificates/readiness/readiness_controller.go#L156
[`trigger_controller.go`]: https://github.com/jetstack/cert-manager/blob/88e85d07253f8d8a99628a2220c7f751a23ef4f1/pkg/controller/certificates/trigger/trigger_controller.go#L183
[`clusterissuers/sync.go`]: https://github.com/jetstack/cert-manager/blob/88e85d07253f8d8a99628a2220c7f751a23ef4f1/pkg/controller/clusterissuers/sync.go#L45
[`issuers/sync.go`]: https://github.com/jetstack/cert-manager/blob/88e85d07253f8d8a99628a2220c7f751a23ef4f1/pkg/controller/issuers/sync.go#L45
[`acmechallenges/sync.go`]: https://github.com/jetstack/cert-manager/blob/88e85d07253f8d8a99628a2220c7f751a23ef4f1/pkg/controller/acmechallenges/sync.go#L68
[`certificaterequests/sync.go`]: https://github.com/jetstack/cert-manager/blob/88e85d07253f8d8a99628a2220c7f751a23ef4f1/pkg/controller/certificaterequests/sync.go#L51

Is there a reason to do the deep copy that early? Early deep copy means less mistakes with mutating objects in the cache, but also less performant (I don't have any proof of this thought, so that's not a good argument).

I'd also choose to do the deep copy "up front" to avoid programming mistakes. But I'd also like us to add a comment that explains to other Kubernetes developers why we choose to deep copy early.

Reply:
I thought about doing it later but it really doesn't matter much to be copying 100's of bytes in a Kubernetes controller like cert-manager.

I will add a comment to this effect.

Reply:
Agreed. I can't really advocate for moving the deep copies deeper in the code paths since I have no proof that it would lead to better performance...